File: Java/FractionAdditionAndSubtraction/FractionAdditionAndSubtraction.java
import java.util.Arrays;

/**
 * (19ms)
 * Created by Phoenix-Z on 2017/5/26.
 */
public class FractionAdditionAndSubtraction {

    static class Fraction {
        int son, mum;

        public Fraction(String fraction) {
            String[] nums = fraction.split("/");
            son = Integer.parseInt(nums[0]);
            mum = Integer.parseInt(nums[1]);
        }

        public void add(Fraction another) {
            int lcm = getLcm(this.mum, another.mum);
            this.son *= lcm / this.mum;
            this.mum = lcm;
            another.son *= lcm / another.mum;
            this.son += another.son;
        }

        public void sub(Fraction another) {
            int lcm = getLcm(this.mum, another.mum);
            this.son *= lcm / this.mum;
            this.mum = lcm;
            another.son *= lcm / another.mum;
            this.son -= another.son;
        }

        public void reduceRedundent() {
            int gcd = getGcd(Math.abs(son), mum);
            if(gcd == 0) {
                son = 0;
                mum = 1;
            } else {
                son /= gcd;
                mum /= gcd;
            }
        }
        public String toString(){
            reduceRedundent();
            return son + "/" + mum;
        }
    }

    public String fractionAddition(String expression) {
        String[] nums = expression.split("[+-]");
        Fraction[] fractions = new Fraction[nums.length];
        if(expression.charAt(0) == '-') {
            nums = Arrays.copyOfRange(nums, 1, nums.length);
            nums[0] = "-" + nums[0];
        }
        for(int i = 0; i < nums.length; i++) {
            fractions[i] = new Fraction(nums[i]);
        }
        Fraction ans = fractions[0];
        int j = 1;
        for(int i = 1; i < expression.length(); i++) {
            if(j >= fractions.length)
                break;
            if(expression.charAt(i) == '+') {
                ans.add(fractions[j++]);
            } else if(expression.charAt(i) == '-') {
                ans.sub(fractions[j++]);
            }
        }
        return ans.toString();
    }

    public static int getGcd(int a, int b) {
        if(a < b) {
            int tmp = a;
            a = b;
            b = tmp;
        }
        if(b == 0 || a % b == 0)
            return b;
        return getGcd(b, a % b);
    }

    public static int getLcm(int a, int b) {
        return a / getGcd(a, b) * b;
    }

    public static void main(String[] args) {
        FractionAdditionAndSubtraction test = new FractionAdditionAndSubtraction();
        System.out.println(test.fractionAddition("-1/2+1/6+1/3"));
    }
}

File: Python/FractionAdditionAndSubtraction/FractionAdditionAndSubtraction.py
class Solution(object):

    def get_gcd(self, a, b):
        if a < b:
            a, b = b, a
        if b == 0 or a % b == 0:
            return b
        return self.get_gcd(b, a % b)

    def get_lcm(self, a, b):
        gcd = self.get_gcd(a, b)
        return a / gcd * b

    def add(self, one, another):
        lcm = self.get_lcm(one[1], another[1])
        one[0] *= lcm / one[1]
        one[1] = lcm
        another[0] *= lcm / another[1]
        one[0] += another[0]
        return one

    def sub(self, one, another):
        lcm = self.get_lcm(one[1], another[1])
        one[0] *= lcm / one[1]
        one[1] = lcm
        another[0] *= lcm / another[1]
        one[0] -= another[0]
        return one

    # (65ms)
    def fractionAddition(self, expression):
        """
        :type expression: str
        :rtype: str
        """
        import re
        nums = re.split('[+-]', expression)
        if expression[0] == '-':
            nums = nums[1:]
            nums[0] = '-' + nums[0]
        fractions = []
        for num in nums:
            fractions.append(map(int, num.split('/')))
        ans, i = fractions[0], 1
        for c in expression[1:]:
            if i >= len(fractions):
                break
            if c == '+':
                ans = self.add(ans, fractions[i])
                i += 1
            elif c == '-':
                ans = self.sub(ans, fractions[i])
                i += 1
        gcd = self.get_gcd(abs(ans[0]), ans[1])
        return '{}/{}'.format(ans[0] / gcd, ans[1] / gcd) if gcd else '0/1'


print Solution().fractionAddition('-4/7-3/4+2/3')
